source/common/websocket/codec.h

```diff
@@ -0,0 +1,143 @@
+class Decoder {
+public:
+  // Decodes the given buffer with WebSocket frame. Drains the input buffer when
+  // decoding succeeded (returns true). If the input is not sufficient to make a
+  // complete WebSocket frame, it will be buffered in the decoder. If a decoding
+  // error happened, the input buffer remains unchanged.
+  // @param input supplies the binary octets wrapped in a WebSocket frame.
+  // @param output supplies the buffer to store the decoded data.
+  // @return bool whether the decoding succeeded or not.
+  absl::optional<std::vector<Frame>> decode(Buffer::Instance& input);
+
+private:
+  // Decodes the given buffer with WebSocket data frames and updates the frame count.
+  // Invokes visitor callbacks for each frame in the following sequence:
+  // --------------------------------------------------------------------------------------------
+  // doDecodeFlagsAndOpcode doDecodeMaskFlagAndLength doDecodeExtendedLength* doDecodeMaskingKey*
+  // doDecodePayload*
+  // --------------------------------------------------------------------------------------------
+  // If doDecodeFlagsAndOpcode returns false, then the decoder aborts.
+  bool inspect(Buffer::Instance&, absl::optional<std::vector<Frame>>&);
+
+  void frameMaskFlag(uint8_t);
+  void frameDataStart(Buffer::Instance&, absl::optional<std::vector<Frame>>&);
+  void frameData(const uint8_t*, uint64_t);
+  void frameDataEnd(Buffer::Instance&, absl::optional<std::vector<Frame>>&);
+
+  bool doDecodeFlagsAndOpcode(uint8_t);
+  void doDecodeMaskFlagAndLength(const uint8_t, Buffer::Instance&,
+                                 absl::optional<std::vector<Frame>>&);
+  void doDecodeExtendedLength(const uint8_t, Buffer::Instance&,
+                              absl::optional<std::vector<Frame>>&);
+  void doDecodeMaskingKey(const uint8_t, Buffer::Instance&, absl::optional<std::vector<Frame>>&);
+  void doDecodePayload(const uint64_t, const uint8_t*&, uint64_t&, Buffer::Instance&,
+                       absl::optional<std::vector<Frame>>&);
+
+  // Current state of the frame that is being processed.
+  enum class State {
+    // Decoding the first byte. Waiting for decoding the final frame flag (1 bit)
+    // and reserved flags (3 bits) and opcode (4 bits) of the WebSocket data frame.
+    FrameHeaderFlagsAndOpcode,
+    // Decoding the second byte. Waiting for decoding the mask flag (1 bit) and
+    // length/length flag (7 bit) of the WebSocket data frame.
+    FrameHeaderMaskFlagAndLength,
+    // Waiting for decoding the extended length of the frame if length read previously
+    // is either 126 or 127. Respectively 2 bytes or 8 bytes will be decoded from the
+    // WebSocket data frame.
+    FrameHeaderExtendedLength,
+    // Waiting for decoding the masking key (4 bytes) only if the mask bit is set.
+    FrameHeaderMaskingKey,
+    // Waiting for decoding the payload (both extension data and application data).
+    FramePayload,
```

Comment:
Let's add an explicit state for `FrameFinished` which will allow the switch statement to have an explicit place for processing the end of the frame.